### PR TITLE
[lodash] Fix GetFieldType indexing for numeric indexes

### PR DESCRIPTION
**Pull request by helgardferreira**
This PR introduces a non-breaking fix to the `GetIndexedField` type used in `GetFieldType` for lodash's `get` function. The current type definition does not work as expected when dealing with deeply nested trees where some of the keys can be numbers instead of strings.

The attached codesandbox link show cases the current behavior as well as a demonstration of the PR's fix working as expected.

**NOTE:** There's a slight caveat however - this fix makes use of a TypeScript feature released in version [4.8](https://devblogs.microsoft.com/typescript/announcing-typescript-4-8/#infer-types-template-strings). Developers using this fix _before_ `4.8` (i.e. the `minimumTypeScriptVersion` configured for `@types/lodash` currently - `4.7.x`) will experience the _old_ behavior whereas anyone using `4.8` and above will be able to benefit from the fix.

I'm not sure how to implement tests that differ based on the TypeScript version so I didn't add any new tests since the output would differ for anything pre-`4.8`. Maybe we should consider upping the `minimumTypeScriptVersion` for `@types/lodash` to `4.8`?

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] [Add or edit tests](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#my-package-teststs) to reflect the change.
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] [Run `pnpm test <package to test>`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#running-tests).

Select one of these and delete the others:

If changing an existing definition:

- [x] Provide a URL to documentation or source code which provides context for the suggested changes: <<url here>>
https://codesandbox.io/p/sandbox/lodash-get-field-type-indexing-rc8xt6?file=%2Fsrc%2Findex.ts%3A48%2C61

### CODE REVIEW

File: types/lodash/common/object.d.ts
Comment (aj-r) [inline types/lodash/common/object.d.ts:1026]: I'm having a hard time following this due to inconsistent indentation. I this this line and below should be indented 1 level deeper.

**Review by aj-r — CHANGES_REQUESTED** [review]
This is quite complex so I don't want to merge without tests. I think it's reasonable to bump `minimumTypescriptVersion` to 4.8.

File: types/lodash/common/object.d.ts
Comment (helgardferreira) [reply types/lodash/common/object.d.ts:1026]: Awesome, thanks for the quick feedback! I'll bump the `minimumTypescriptVersion` to `4.8` and write some unit tests for the change.

I'll also adjust the indentation level (for some reason prettier & eslint is acting up for me in the cloned repo).

**Comment by helgardferreira** [comment]
@aj-r So I managed to bump the `minimumTypescriptVersion` to `4.8` for `lodash` but as you'll see in the last few commits I had to then bump the `minimumTypescriptVersion` for _several_ other packages as well (~300). I hope this is ok? 😅

If not, please let me know. Also, I made a small codemod you can find [here](https://github.com/helgardferreira/codemods) to batch update the `minimumTypescriptVersion`s.

**Comment by aj-r** [comment]
Yikes, in theory we'd need approval from maintainers of the other packages in order to bump the version. I wonder if any DT maintainers know a way to write a test only for TS4.8 while keeping the `minimumTypescriptVersion` at 4.7?

File: types/lodash/common/object.d.ts
Comment (rbuckton) [inline types/lodash/common/object.d.ts:1025]: Does it matter that `1.23` and `.123` will work for this? It's fair if it doesn't matter since the previous code did not check against this, but it's possible to verify the result is an integer using this:
```suggestion
        : [K, K] extends [`${bigint}`, `${infer U extends number}`]
```

**Comment by jakebailey** [comment]
> Yikes, in theory we'd need approval from maintainers of the other packages in order to bump the version. I wonder if any DT maintainers know a way to write a test only for TS4.8 while keeping the `minimumTypescriptVersion` at 4.7?

There isn't a way currently, though it is possible that we could extend https://github.com/microsoft/DefinitelyTyped-tools/pull/973 to filter specific tsconfigs by TS versions.

That being said TS 4.7 will go out of support once 5.5 is out, so waiting a bit may also "fix" the problem.

**Comment by aj-r** [comment]
> That being said TS 4.7 will go out of support once 5.5 is out, so waiting a bit may also "fix" the problem.

Sounds good, let's just wait for 5.5 to be released then.

**Comment by helgardferreira** [comment]
> > That being said TS 4.7 will go out of support once 5.5 is out, so waiting a bit may also "fix" the problem.
> 
> Sounds good, let's just wait for 5.5 to be released then.

Ok, awesome. Sounds good to me 🙂

**Comment by sandersn** [comment]
@typescript-bot this PR should stay open till 5.5 is out and 4.7 support is removed from DT.

**Comment by jakebailey** [comment]
Converting the PR to a draft will make the bot stop complaining.

**Review by sheetalkamat — DISMISSED** [review]
Please convert to draft and reopen after 5.5 is released

**Comment by jakebailey** [comment]
4.7 is now out of support and no longer tested, so this PR should be safe to revive with the most recent commit reverted.

**Comment by helgardferreira** [comment]
@jakebailey Legend! Thanks for the clean up 😁

**Review by andrewbranch — APPROVED** [review]

**Comment by aj-r** [comment]
Hey looks like #69696 just got merged. Does that solve the same use case as this one?

**Pull request closed without merging**
